kore/src/Kore/Step/Implication.hs

```diff
@@ -0,0 +1,463 @@
+assertRefreshed implication'@Implication{ existentials } =
+    assert (isFreeOf implication' (Set.fromList $ inject <$> existentials))
+
+{- | Refresh the 'existentials' of the 'Implication'.
+ -}
+refreshExistentials :: Implication modality -> Implication modality
+refreshExistentials = snd . refreshRule mempty
+
+instance UnifyingRule (Implication modality) where
+    type UnifyingRuleVariable (Implication modality) = RewritingVariableName
+
+    matchingPattern implication'@(Implication _ _ _ _ _) =
+        Pattern.term left
+      where
+        Implication { left } = implication'
+
+    precondition implication'@(Implication _ _ _ _ _) =
+        Condition.toPredicate . Pattern.withoutTerm $ left
+      where
+        Implication { left } = implication'
+
+    refreshRule stale implication'@(Implication _ _ _ _ _) =
+        do
+            let variables = freeVariables implication' & FreeVariables.toSet
+            renaming <- refreshVariables' variables
+            refreshedImplication <- renameExistentials
+            let fullyRefreshedImplication = substitute
+                    (TermLike.mkVar <$> renaming)
+                    refreshedImplication
+            -- Only return the renaming of free variables.
+            -- Renaming the bound variables is invisible from outside.
+            pure (renaming, fullyRefreshedImplication)
+        & flip evalState (FreeVariables.toNames stale)
+      where
+        refreshVariables' variables = do
+            staleNames <- State.get
+            let renaming = refreshVariables staleNames variables
+                staleNames' = Set.map variableName variables
+                staleNames'' =
+                    Map.elems renaming
+                    & foldMap FreeVariables.freeVariable
+                    & FreeVariables.toNames
+            State.put (staleNames <> staleNames' <> staleNames'')
+            pure renaming
+
+        Implication { right, existentials } = implication'
+
+        renameExistentials = do
+            let existentials' = Set.fromList (inject <$> existentials)
+            rename <- refreshVariables' existentials'
+            let renameVariable var =
+                    maybe var TermLike.expectElementVariable
+                    $ Map.lookup
+                        (SomeVariableNameElement (variableName var))
+                        rename
+                subst = TermLike.mkVar <$> rename
+            pure $ implication'
+                { right = OrPattern.substitute subst right
+                , existentials = renameVariable <$> existentials
+                }
+
+resetConfigVariables :: Implication modality  -> Implication modality
+resetConfigVariables implication'@(Implication _ _ _ _ _) =
+    implication'
+        { left =
+            Pattern.mapVariables resetConfigVariable left
+        , right =
+            OrPattern.map
+                (Pattern.mapVariables resetConfigVariable)
```

Comment:
Let's introduce a `mapVariables` function in `Kore.Internal.OrPattern` which combines `OrPattern.map` and `Pattern.mapVariables`. I'm sure this combination is used elsewhere, too.